### PR TITLE
ENH: Add framework for property model parameter generation

### LINKED ISSUE #5: Support parameter selection for custom single phase models

### PR DESCRIPTION
**Pull request by bocklund**
This is a significant feature improvement that introduces the concept of a `ModelFittingDescription` made up of `FittingStep` objects to tell ESPEI how to generate model parameters (with a capability to support generating parameters for custom models). In addition, these changes include:

- Significant refactoring and simplification of `espei.paramselect`, the main module for parameter selection:
    - Most of the code that was duplicated for fitting endmembers, binary interactions, and ternary interactions was unified. 
    - The `fit_ternary_interactions` function was completely removed and binary and ternary interactions now run the exact same code path.
    - The most significant changes are in `fit_parameters`, which now relies on dependency injection of a fitting description to know what fitting steps to take (we try to not make any assumptions about what parameters or data types are being fit) and this function is more streamlined to 1) select relevant data from datasets, 2) get the RHS (`b` in `Ax=b`) of the linear problem from the fitting step, 3) build candidate models (build a collection feature matrices, `A`), 4) select the best model from the candidates, and 5) insert the parameter and coefficients for the best model into the database
- Refactoring to remove `espei.parameter_selection.utils`. Some of the functionality got moved to be in fitting steps, and `get_sample_condition_dicts` was moved to `espei.error_functions.non_equilibrium_thermochemical_error`, with `get_prop_samples` that it is used in conjunction with. Trying to have fewer "utils" modules. 
- Tests to verify that molar volume and custom property models (elastic) work against various edge cases
- New docs page with a tutorial for using a custom model and fitting description for fitting elastic constants with a [repository](https://github.com/bocklund/espei-elastic-parameter-generation) containing the code
- Small tweak to follow links in dataset recursive glob search. This is useful for organizing datasets for different runs while having one single source of truth for the data

This PR relies on https://github.com/pycalphad/pycalphad/pull/501 that includes fixes for molar volume models. Tests will probably fail until that gets into the development branch.

Closes #5

### CODE REVIEW

**Comment by bocklund** [comment]
@nury12n if you could review :)

File: espei/parameter_selection/fitting_steps.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+
+    @classmethod
+    def get_response_vector(cls, fixed_model: Model, fixed_portions: [symengine.Basic], data: [Dataset], sample_condition_dicts: [Dict[str, Any]]) -> ArrayLike:  # np.float_
+        mole_atoms_per_mole_formula_unit = fixed_model._site_ratio_normalization
+        # Define site fraction symbols that will be reused
+        phase_name = fixed_model.phase_name
+
+        # Construct flattened list of site fractions corresponding to the ravelled data (from shift_reference_state)
+        site_fractions = []
```
Comment (nury12n) [inline espei/parameter_selection/fitting_steps.py:248]: I noticed lines 248-253 + 260-270 is pretty much the same as lines 152-169 where it replaces the symbols for site fractions and state variables with the actual values. Do you think it makes sense for these lines to be a separate function in FittingStep which could make the API for implementing a subclass of FittingStep a bit easier?

I'm not sure if it's very necessary to do so since a user subclassing FittingStep would likely being looking at StepHM as a reference in any case. And having these lines as a separate function may make it less clear that they are necessary if symengine symbols are introduced to data_qtys.

File: espei/parameter_selection/fitting_steps.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+            sf.update({x: 0. for x in missing_variables})
+            sf.update(cond_dict)
+        # also replace with database symbols in case we did higher order fitting
+        rhs = [fixed_model.symbol_replace(symengine.S(value_with_symbols).xreplace(sf), fixed_model._symbols).evalf() for value_with_symbols, sf in zip(rhs, site_fractions)]
+        # cast to float, confirming that these are concrete values with no sybolics
+        rhs = np.asarray(rhs, dtype=np.float_)
+        return rhs
+
+
+class StepHM(FittingStep):
```
Comment (nury12n) [inline espei/parameter_selection/fitting_steps.py:173]: Are there other properties like HM, SM or CPM where the data is linear to the feature and the feature is just a transformation of expr (and expr corresponds to the model attribute)? If there is, do you think it'll make sense to have something similar to AbstractLinearFittingStep (maybe AbstractNonLinearFittingStep) where the user just has to implement transform_feature and possibly transform_data?

I don't know if doing this could obfuscate things and cause the user to think that anything can be fitted with either AbstractLinearFittingStep or AbstractNonLinearFittingStep. When I was fitting directly to the mobility models (rather than fitting to the RK polynomial portion of the mobility), I realized I had to do two different transformation (expr -> linear expression to dataset and model attribute -> linear expression to dataset since the model attribute and expr was different) which I probably only figured out after understanding what StepHM was doing. So it might just be better to leave it so if the user is fitting to something more complicated, they would be able to look at StepHM as a reference to figure out what to do for non-linear properties.

**Review by nury12n — APPROVED** [review]
Added some comments regarding the fitting steps, mainly thoughts I had when working with the FittingStep and AbstractLinearFittingStep classes, but I think it'd probably be better to leave as is.

I gave some thought about the diffusing species and I think it'll make more sense to consider them after there's some support in kawin for mobility fitting.

**Comment by bocklund** [comment]
> Added some comments regarding the fitting steps, mainly thoughts I had when working with the FittingStep and AbstractLinearFittingStep classes, but I think it'd probably be better to leave as is.
> 
> I gave some thought about the diffusing species and I think it'll make more sense to consider them after there's some support in kawin for mobility fitting.

Thanks for reviewing! Since PyCalphad supports diffusion species, it's something that could be in scope for ESPEI, but my guess is that there are assumptions elsewhere in the parameter generation code that would need to be broken to be able to handle that extra dimension that parameters can also be a function of diffusing species. At least, I think having these `FittingStep` and `FittingDescription` abstractions are compatible with what would be needed to fit parameters with diffusion species and it's mostly a book-keeping and dataset selection operations in `espei.paramselect.phase_fit` and `espei.paramselect.fit_parameters`

**Comment by bocklund** [comment]
Tests are passing on pycalphad development versions and not release versions (as expected from https://github.com/pycalphad/pycalphad/pull/501). Merging
